bip-XXXX.mediawiki

```diff
@@ -0,0 +1,65 @@
+<pre>
+  BIP: XXXX
+  Layer: Consensus (hard fork)
+  Title: Limit number of transactions which contain an OP_RETURN opcode to one per block
+  Author: Christopher Gilliard <christopher.gilliard@gmail.com>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
+  Status: Draft
+  Type: Standards Track
+  Created: 2021-04-16
+  License: BSD-2-Clause
+</pre>
+
+==Abstract==
+
+This BIP proposes limiting the number of transactions which contain an OP_RETURN opcode to one per block.
+
+==Motivation==
+
+For background, the OP_RETURN opcode [<a href="#references">1</a>], is used in Bitcoin to store arbitrary additional data into a block which is then permanently stored as part of the blockchain. Today there is a limitation of one OP_RETURN per transaction and each OP_RETURN is limited to 80 bytes of data. With very limited block space, it is essential to optimize the data stored in a block. Since a Merkle Tree [<a href="#references">2</a>] or other similar data structures can be used to timestamp multiple documents by storing a single 32 byte checksum of the top of a Merkle Tree in a block, there is no reason to allow more than one OP_RETURN per block. If only a single OP_RETURN is allowed, per block, users will be required to coordinate in order to store individual checksums into this data structure, but this can be achieved by developing a layer two time-stamping protocol. This layer two time-stamping protocol will allow users to store thousands or even millions of checksums into a single block and at the same time limit the total amount of data in the block to 80 bytes. A future BIP proposing such a layer two protocol will be forthcoming. Use cases other than time-stamping will be harder to achieve with this proposal, and that is part of the goal of this BIP since those other use cases should be discouraged by the rules of the protocol.
+
+==Specification==
+
+After deployment (see the Deployment section for details), blocks that contain more than one OP_RETURN opcode will be rejected by full nodes. Additionally, new logic will be needed for the transaction pool because miners must select a single OP_RETURN transaction to include in the block. Ultimately, miners will run a different piece of code that builds the tree structure required to timestamp data and include their own OP_RETURN transaction in blocks they mine. A future BIP outlining the layer two protocol will describe an incentive structure so that miners will benefit from building the trees in the layer two protocol. The entire tree will be downloadable via a decentralized storage mechanism that will be described in a forthcoming BIP as well.
```

Comment:
What if scripts wish to use `OP_RETURN` conditionally for some programmic purpose? (Not required to fix for BIP assignment/merge, just something IMO should be clarified)

Reply:
Do you know of any examples of this? What would the use case be?

Reply:
Nothing comes to mind immediately, but presumably it would be handy within an `OP_IF` block.